src/crimson/os/seastore/segment_manager/block.cc

```diff
@@ -0,0 +1,416 @@
+      return do_read(
+	device,
+	0,
+	bp
+      ).safe_then([=, &bp] {
+	  bufferlist bl;
+	  bl.push_back(bp);
+	  block_sm_superblock_t ret;
+	  auto bliter = bl.cbegin();
+	  decode(ret, bliter);
+	  return BlockSegmentManager::access_ertr::future<block_sm_superblock_t>(
+	    BlockSegmentManager::access_ertr::ready_future_marker{},
+	    ret);
+      });
+    });
+}
+
+BlockSegment::BlockSegment(
+  BlockSegmentManager &manager, segment_id_t id)
+  : manager(manager), id(id) {}
+
+segment_off_t BlockSegment::get_write_capacity() const
+{
+  return manager.get_segment_size();
+}
+
+Segment::close_ertr::future<> BlockSegment::close()
+{
+  manager.segment_close(id);
+  return close_ertr::now();
+}
+
+Segment::write_ertr::future<> BlockSegment::write(
+  segment_off_t offset, ceph::bufferlist bl)
+{
+  if (offset < write_pointer || offset % manager.superblock.block_size != 0)
+    return crimson::ct_error::invarg::make();
+
+  if (offset + bl.length() > manager.superblock.segment_size)
+    return crimson::ct_error::enospc::make();
+
+  return manager.segment_write({id, offset}, bl);
+}
+
+Segment::close_ertr::future<> BlockSegmentManager::segment_close(segment_id_t id)
+{
+  assert(tracker);
+  tracker->set(id, segment_state_t::CLOSED);
+  return tracker->write_out(device, superblock.tracker_offset);
+}
+
+Segment::write_ertr::future<> BlockSegmentManager::segment_write(
+  paddr_t addr,
+  ceph::bufferlist bl,
+  bool ignore_check)
+{
+  assert((bl.length() % superblock.block_size) == 0);
+  logger().debug(
+    "segment_write to segment {} at offset {}, physical offset {}, len {}",
+    addr.segment,
+    addr.offset,
+    get_offset(addr),
+    bl.length());
+
+  
+  // TODO send an iovec and avoid the copy -- bl should have aligned
+  // constituent buffers and they will remain unmodified until the write
+  // completes
+  return seastar::do_with(
+    bufferptr(ceph::buffer::create_page_aligned(bl.length())),
+    [&](auto &bp) {
+      auto iter = bl.cbegin();
+      iter.copy(bl.length(), bp.c_str());
+      return do_write(device, get_offset(addr), bp); // TODO flushes
+    });
+}
+
+BlockSegmentManager::~BlockSegmentManager()
+{
+}
+
+BlockSegmentManager::mount_ret BlockSegmentManager::mount(mount_config_t config)
+{
+  return open_device(
+    config.path, seastar::open_flags::rw
+  ).safe_then([=](auto p) {
+    device = std::move(p.first);
+    auto sd = p.second;
+    return read_superblock(device, sd);
+  }).safe_then([=](auto sb) {
+    superblock = sb;
+    tracker = std::make_unique<SegmentStateTracker>(
+      superblock.segments,
+      superblock.block_size);
+    return tracker->read_in(
+      device,
+      superblock.tracker_offset
+    ).safe_then([this] {
+      for (segment_id_t i = 0; i < tracker->get_capacity(); ++i) {
+	if (tracker->get(i) == segment_state_t::OPEN) {
+	  tracker->set(i, segment_state_t::CLOSED);
+	}
+      }
+      return tracker->write_out(device, superblock.tracker_offset);
```

Comment:
i am confused. does this imply that we are able to rewrite a block in-place?

Reply:
Yes, BlockSegmentManager uses a conventional device to implement the SegmentManager interface.  This portion simply maintains a control block on the disk recording the state of each segment.  I expect that zns devices will get an analogous ZNSSegmentManager implementation that use the zns library to directly query these states.

Reply:
makes sense to me. by then we will be able to issue commands to query the segment states. could you note this down in `block.h` ? a very short note would suffice.